### ISSUE TITLE
Only Exact Match if Phrases are Used

### ISSUE THREAD

**Issue by freedombird9** [topheman/twitter-stream-channels#5]
I noticed that this module can only do exact match if I give it a phrase, such as 'Bay Area Weather'. I wonder if it's possible to imitate the behavior of twitter streaming API's original track. So the above phrase should be equivalent to 'Bay ADN Area AND Weather', which means any tweet that contains all of them would be filtered out, no matter the order of appearance or if they're consecutive or not.

**Comment by topheman** [comment]
You're right, this module **only does exact matching** and it won't change. The main reasons being:

* performance (it could be postprocessing a lot of tweets in real time)
* being able to match what was found in the tweet with what you were searching for (ex: I'm searching for "big apple", it tells me it found "big apple" in a tweet, I can highlight it in the text of the tweet easily)

What you describe can be done but I don't think you want to use twitter-stream-channels (or any other module of that kind), you'll want to directly connect to the twitter stream API, and post-process yourself the tweets, which will allow you to **choose exactly your matching strategy**.

If you do it in realtime, your matching strategy will have to be very optimized for your specific use case. If you don't do "real" realtime (like you populate a db and display the results with some delay), you'll be able to use advanced matching (like db querying, even use some elasticsearch like engine).

**Comment by freedombird9** [comment]
Thanks for your prompt reply. I see your point. In my case, I think it would make more sense to do a kind of topic match. For the above example, I just need to give the words 'Bay Area Weather', then any tweet that contains them would be delivered to me, and I can choose what I want from the results.

In such cases, it is hard to come up with an exact phrase. I just want a topic (bay area weather). I would need to compose a lot of different combinations in order not to miss any tweet I want if I use exact match.

And I wonder if you're using the sampling API? Since I think Twitter doesn't allow direct connections to the stream. I think it limits our bandwidth. Traffic directed to us cannot exceed a certain percentage of the stream. So maybe a 10% sampling and then post-processing? Thanks.

**Comment by topheman** [comment]
> In such cases, it is hard to come up with an exact phrase. I just want a topic (bay area weather). I would need to compose a lot of different combinations in order not to miss any tweet I want if I use exact match.

That's exactly what I meant.

> And I wonder if you're using the sampling API?

No I don't.

> I think it limits our bandwidth. Traffic directed to us cannot exceed a certain percentage of the stream. So maybe a 10% sampling and then post-processing?

If I remember correctly, there is no limitation on that matter. You can connect:

* To a filtered steam (like I do)
* To a sample (which represent 1% or something like that of all the realtime tweets)
* To the firehose:  directly to **all** the tweets from twitter

The limit is your bandwidth AND your computing power (before even starting to post-process all those tweets, you'll need to consume them - if you can't follow up, twitter disconnects you). So, I'd say your best option remains to:

* connect to a filtered stream (that way you already filter what's coming at you, in your case: all the tweets with 'Bay' in it for example)
* post process with an advanced regex that can match the specific patterns you target

**Comment by freedombird9** [comment]
I see. Thank you very much. It really helps a lot!

**Comment by topheman** [comment]
Glad i could help you @freedombird9

**Comment by mrwensveen** [comment]
A bit late to the party, but I've looked at this for a bit. I think you should be able to modify `stream.channelsKeywordsLowerCasedRegExp` to get what you want. `stream.channelsKeywordsLowerCasedRegExp[channel]` is an array of Regexes that represent your keywords when the tweets are being post-processed.

For example, if the current keyword is `"bay area weather"`, you can replace it with `"(?=.*\bbay\b)(?=.*\barea\b)(?=.*\bweather\b).*"`. This should match sentences like "the weather in the bay area is fine today" (all matches are done lower-cased). This shouldn't impact performance much since it's still just a regular expression match being done.

Untested, so YMMV.

@topheman I can try to make this an option to do this automatically. Would you be willing to accept pull requests?

BTW: This would also reflect the stream api better (https://developer.twitter.com/en/docs/tweets/filter-realtime/guides/basic-stream-parameters.html#track)

**Comment by mrwensveen** [comment]
Hi. I've created a working solution.
After starting the stream with `streamChannels`, I parse all keywords (lower-case, regex escape, same as before), then I call `matchAllRegExp` (see below) before joining them with '|' and creating a RegExp object: 
The new RegExp object can then be set for the channel with `stream.getChannelsKeywordsLowerCasedRegExp()[channel] = regex;`

This works perfectly and without any visible performance impact. While I'm able to do this in my own code (no fork necessary), I did need to duplicate the `regExpEscape` function and some parts of `preprocessKeywords`. It would be neat if this code could be integrated upstream, possibly as an option (e.g. `matchExact` (true by default, old behavior)).

```javascript
  matchAllRegExp = function(regex) {
    // Turn 'bay area weather' into '(?=.*\bbay\b)(?=.*\barea\b)(?=.*\bweather\b).*'
    return regex.indexOf(' ') >= 0 ?
      '(?=.*\\b' + regex.split(' ').join('\\b)(?=.*\\b') + '\\b).*' :
      regex;
  }
```
Anyway, thanks for the cool library.
